src/iceberg/util/decimal.cc

```diff
@@ -0,0 +1,1044 @@
+  // the input represents 9876543210123456789, then segments will be [123456789,
+  // 876543210, 9].
+  // The max number of segments needed = ceil(kNumBits * log(2) / log(1e9))
+  // = ceil(kNumBits / 29.897352854) <= ceil(kNumBits / 29).
+  std::array<uint32_t, (kNumBits + 28) / 29> segments;
+  size_t num_segments = 0;
+
+  while (uval > 0) {
+    // Compute remainder = uval % 1e9 and uval = uval / 1e9.
+    auto remainder = static_cast<uint32_t>(uval % k1e9);
+    uval /= k1e9;
+    segments[num_segments++] = remainder;
+  }
+
+  std::ostringstream oss;
+  if (negative) {
+    oss << '-';
+  }
+
+  // First segment is formatted as-is.
+  oss << segments[num_segments - 1];
+
+  // Remaining segments are formatted with leading zeros to fill 9 digits. e.g. 123 is
+  // formatted as "000000123"
+  for (size_t i = num_segments - 1; i-- > 0;) {
+    oss << std::setw(9) << std::setfill('0') << segments[i];
+  }
+
+  return oss.str();
+}
+
+Result<Decimal> Decimal::FromString(std::string_view str, int32_t* precision,
+                                    int32_t* scale) {
+  if (str.empty()) {
+    return InvalidArgument("Empty string is not a valid Decimal");
+  }
+  DecimalComponents dec;
+  if (!ParseDecimalComponents(str, &dec)) {
+    return InvalidArgument("Invalid decimal string '{}'", str);
+  }
+
+  // Count number of significant digits (without leading zeros)
+  size_t first_non_zero = dec.while_digits.find_first_not_of('0');
+  size_t significant_digits = dec.fractional_digits.size();
+  if (first_non_zero != std::string_view::npos) {
+    significant_digits += dec.while_digits.size() - first_non_zero;
+  }
+
+  auto parsed_precision = static_cast<int32_t>(significant_digits);
+
+  int32_t parsed_scale = 0;
+  if (dec.has_exponent) {
+    auto adjusted_exponent = dec.exponent;
+    parsed_scale = static_cast<int32_t>(dec.fractional_digits.size()) - adjusted_exponent;
+  } else {
+    parsed_scale = static_cast<int32_t>(dec.fractional_digits.size());
+  }
+
+  uint128_t value = 0;
+  ShiftAndAdd(dec.while_digits, value);
+  ShiftAndAdd(dec.fractional_digits, value);
+  Decimal result(static_cast<int128_t>(value));
+
+  if (dec.sign == '-') {
+    result.Negate();
+  }
+
+  if (parsed_scale < 0) {
+    // For the scale to 0, to avoid negative scales (due to compatibility issues with
+    // external systems such as databases)
+    if (parsed_scale < -kMaxScale) {
+      return InvalidArgument("scale must be in the range [-{}, {}], was {}", kMaxScale,
+                             kMaxScale, parsed_scale);
+    }
+
+    result *= kDecimal128PowersOfTen[-parsed_scale];
+    parsed_precision -= parsed_scale;
+    parsed_scale = 0;
+  }
+
+  if (precision != nullptr) {
+    *precision = parsed_precision;
+  }
+  if (scale != nullptr) {
+    *scale = parsed_scale;
+  }
+
+  return result;
+}
+
+namespace {
+
+constexpr float kFloatInf = std::numeric_limits<float>::infinity();
```

Comment:
As I have commented in the decimal.h, if we decide not to support floating-point conversion, we can avoid maintaining a large portion of code starting from here.